Add insert cell above and insert cell below commands

mattbui opened the pull request:
I added 2 commands: `IPythonCellInsertAbove` and `IPythonCellInsertBelow` insert a new cell above/below current cell.

The cell tag inserted can be specified via `g:ipython_cell_insert_tag` (defaults to `g:ipython_cell_tag[0]`).

Please take a look, and thanks for the amazing extension.

![insert_comamnds](https://user-images.githubusercontent.com/46804938/118133550-2ad30d00-b42b-11eb-8d31-e174f52c6772.gif)

mattbui committed:
insert cell above & below functions

mattbui committed:
fix some corner cases

mattbui committed:
fix typo

mattbui committed:
fix typo

mattbui committed:
insert_above jump to start_row instead of using jump_prev_cell()

mattbui committed:
rename to have ipythoncell prefix

hanschen commented:
Thanks for the pull request, I'll try to review it this weekend.

hanschen commented:
I've tried it out and it's really nice, great job! Some small questions before I merge this:

1. If the file looks like this (`█` is my cursor position):
   ```
   # %% 
   print("hi")█
   ```
   and I issue `IPythonCellInsertAbove`, I get
   ```
   █

   # %%
   
   # %% 
   print("hi")█
   ```

   Wouldn't it make more sense if it resulted in
   ```
   # %%
   █
   
   # %% 
   print("hi")█
   ```
   instead?

2. I think some people would like to have the cursor end up after the cell tag rather than on a new line. With this in mind, maybe the default cell insert tag could be e.g. `# %% ` (notice the space after `%%`), and if you want to have the current behavior, you could specify `let g:ipython_cell_insert_tag = '# %%\n'` or something similar.

3. Related to the previous point, I would make the default cell tag explicit rather than `g:ipython_cell_tag[0]`.

mattbui committed:
insert header if the first line does not contains cell header when insert above

mattbui committed:
keep the cursor in the header line, change default insert tag

mattbui committed:
padding extra line for insert below

mattbui committed:
fix wrong padding in insert below

mattbui commented:
@hanschen Thanks for taking a look!

1. I addressed this issue by always adding a header for the cell that contains the first line but doesn't have a header and adding an above header afterward. The behavior will look like this:
Before:
```python
# %% 
print("hi")█
```
-> Insert above:
```python
# %%█

# %% 
print("hi")
```

Before:
```python
print("hi")█
```
-> Insert above:
```python
# %%█

# %% 
print("hi")
```

2. & 3. I changed the default insert tag to `# %% ` and keep the cursor at the header after insert, illustrated above.

hanschen commented:
Almost there! :) 

What do you think about making it behave like this:

```
print("hi")█
```

-> Current insert above:

```
# %% █

# %%
print("hi")
```
-> Suggested insert above:

```
# %% █
print("hi")
```

This way you can name the current cell and issue the `IPythonCellInsertAbove` again if you want to insert another empty cell.

I'm also thinking about the newline after the cell tag. How about not adding a newline if the new cell is the last cell? That would avoid e.g. `flake8` complaining about `blank line at end of file`.

For example:
```
# %% Cell 1

print("hi")█
```

-> Current insert below:
```
# %% Cell 1

print("hi")

# %% █

```

-> Suggested insert below:
```
# %% Cell 1

print("hi")

# %% █
```

For insert above/below between cells, I like the current behavior of adding the cell tag and a newline.

mattbui committed:
insert above only create header when first line not a header

mattbui committed:
avoid adding newline if cursor at end of file

mattbui commented:
@hanschen the 2 new commits addressed the problems you mentioned. Thanks for the suggestions 😄 .

hanschen commented:
Awesome! I'm going to merge the pull request now, thanks so much for your contribution. :smiley:
